presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -342,4 +351,200 @@ public void close()
     {
         Optional<ConnectorOutputMetadata> finishTable(Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics);
     }
+
+    public interface PartitionCommitter
+    {
+        void commitPartition(int partitionId, Collection<Slice> fragments);
+    }
+
+    private static class LifespanStageStateTracker
+    {
+        private final Map<LifespanStage, LifespanStageState> unrecoverableLifespanStageStates = new HashMap<>();
+        private final Map<LifespanStage, Map<Integer, LifespanStageState>> unfinishedRecoverableLifespanStageStates = new HashMap<>();
+        private final Map<LifespanStage, LifespanStageState> finishedRecoverableLifespanStages = new HashMap<>();
+
+        private final PartitionCommitter partitionCommitter;
+        private final JsonCodec<TableCommitContext> tableCommitContextCodec;
+
+        public LifespanStageStateTracker(PartitionCommitter partitionCommitter, JsonCodec<TableCommitContext> tableCommitContextCodec)
+        {
+            this.partitionCommitter = requireNonNull(partitionCommitter, "partitionCommitter is null");
+            this.tableCommitContextCodec = requireNonNull(tableCommitContextCodec, "tableCommitContextCodec is null");
+        }
+
+        public void update(Page page)
+        {
+            TableCommitContext tableCommitContext = getTableCommitContext(page);
+            LifespanStage lifespanStage = LifespanStage.fromTableCommitContext(tableCommitContext);
+            if (finishedRecoverableLifespanStages.containsKey(lifespanStage)) {
+                return;
+            }
+
+            switch (tableCommitContext.getCommitGranularity()) {
+                case TABLE:
+                    unrecoverableLifespanStageStates.putIfAbsent(lifespanStage, new LifespanStageState());
```

Comment:
???  Then the granularity shouldn't using `TABLE` and `PARTITION` right?

Maybe `TASK` and `LIFESPAN` ?

Reply:
Per offline discussion, we might want to move all these decision into planner, so we don't need to pass the ambiguous `CommitGranularity` around. I'll work on this and see if it works.

Reply:
Actually that is not as easy and might deserve might thoughts given all the weird possible plan shapes and how `PlanFragmenter` works today.

I'll keep the current method, but change `CommitGranularity` to `lifespanCommitRequired` boolean flag. The `CommitGranularity` confusion comes in two flavors from `UNION` case:

- if some child stages are recoverable and the others are not, it is confusing for unrecoverable stages to claim `CommitGranularity` is `TABLE` (because other recoverable stages writing to the same table has `CommitGranularity` as `PARTITION`).
- Using enum implies that we should do either table commit or lifespan commit, which is not true. We always do table commit, and might do lifespan commit if applicable.